crates/scmi/src/scmi.rs

```diff
@@ -0,0 +1,1377 @@
+// SPDX-FileCopyrightText: Red Hat, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementation of SCMI and some of its protocols.
+//!
+//! This module implements SCMI infrastructure and some of the SCMI protocols.
+//! See [make_handlers] how to add support for another SCMI protocol or to add
+//! more functionality to an already implemented SCMI protocol.
+//!
+//! If you want to add new devices (e.g. SCMI bindings to some kinds of host
+//! devices), see [crate::devices] modules.
+
+use std::{
+    cmp::min,
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+
+use itertools::Itertools;
+use log::{debug, error, info, warn};
+use thiserror::Error as ThisError;
+
+pub(crate) type MessageHeader = u32;
+
+pub(crate) const MAX_SIMPLE_STRING_LENGTH: usize = 16; // incl. NULL terminator
+
+/// Wrapper around SCMI values of the basic types SCMI defines.
+///
+/// Everything communicating to/from SCMI must be composed of them.
+// SCMI specification talks about Le32 parameter and return values.
+// VirtIO SCMI specification talks about u8 SCMI values.
+// Let's stick with SCMI specification for implementation simplicity.
+#[derive(Clone, Debug, PartialEq)]
+pub(crate) enum MessageValue {
+    Signed(i32),
+    Unsigned(u32),
+    String(String, usize), // string, expected characters
+}
+
+impl MessageValue {
+    pub(crate) fn get_unsigned(&self) -> u32 {
+        match self {
+            MessageValue::Unsigned(value) => *value,
+            _ => panic!("Wrong parameter"),
+        }
+    }
+}
+
+pub(crate) type MessageValues = Vec<MessageValue>;
+
+/// Enumeration of SCMI message types, mapped to the corresponding SCMI codes.
+///
+/// The only one we currently support is [MessageType::Command].
+#[derive(Debug, PartialEq)]
+enum MessageType {
+    // 4-bit unsigned integer
+    Command,     // 0
+    Unsupported, // anything else
+}
+pub(crate) type MessageId = u8;
+pub(crate) type ProtocolId = u8;
+type NParameters = u8;
+
+/// Mapping of return values to SCMI return status codes.
+#[derive(Clone, Copy)]
+#[allow(dead_code)]
```

Comment:
is it really dead code?

Reply:
Could use a comment i guess

Reply:
Done.

Reply:
Done.

Reply:
A good pattern is to put all logic in a lib.rs crate, and have the binary use it in main.rs. This way, everything is public and there's no dead code warnings (which obviously has cons too)

Reply:
Well, here, I think it's good to have dead code warnings and silence them only where needed.